### PR TITLE
Start cron control runner on php containers

### PR DESCRIPTION
**Pull request by WPprodigy**
Currently cron is not processed on dev-envs, as we set the `WPCOM_VIP_LOAD_CRON_CONTROL_LOCALLY` constant but never kick off the runner.

So this adds the cron runner to php containers in a background process. Using pretty conservative intervals, will only check for site changes every 120seconds and each site will get events to run every 60 seconds.

I tried to do this as a dev-tools script and w/ lando run, run_as_root, and various lando events. In each case, for an unknown reason, the background process would always disappear during dev-env's first boot-up. Subsequent stop/starts would work with some of them, but alas. This should be quite okay though, as we can pass lando-file / template-specific variables to the `run.sh` if needed.

## Testing

In the lando file, replace the php image with a local copy, like this:

```
  php:
    type: compose
    services:
      build:
        context: /Users/caleb/code/vip/platform/vip-container-images
        dockerfile: /Users/caleb/code/vip/platform/vip-container-images/php-fpm/Dockerfile
```

Then do `dev-env create`, `dev-env start`, etc. Once running, should see `Starting cron runner` in the php container logs. But then for further checking can `exec` into the container and see the process running with `ps`. Logs for the runner can be seen: `tail -f /tmp/cron-runner.log`. Logs should pretty quiet unless real errors happen. if you want to see more runner-related logs, you'll need to pass `-debug` to the cron-control-runner CLI.

### CODE REVIEW

**Comment by WPprodigy** [comment]
Hmm, I'm not totally sure what to do about the build fail. Seems valid, according to https://docs.docker.com/buildx/working-with-buildx/

**Comment by pschoffer** [comment]
@WPprodigy I tested it and it works. Great job!

Few notes:
* Can we redirect logs to `/wp/log/cron-runner.log`? If we do that they will show up on host computer under logs as that folder is already mapped.
* Running it in`run.sh` means 2 things - running as root and running potentially before wp install. I think it would be better to avoid those two.
* As you say it doesnt work if you just add it to the setup.sh. I tested it and it seems due to the lack of permissions.
When I add this at the end of setup in php container:
```
# Kick off cron runner in a background process. Initial sleep is to help give WP time to be setup.                      
echo "Starting cron runner"                                                                                             
nohup bash -c 'sleep 60; /usr/local/bin/cron-control-runner \                                   
    -wp-path=/wp -wp-cli-path=/usr/local/bin/wp \                                                                       
    -get-sites-interval=120s -get-events-interval=60s \                                                                 
    -get-events-parallelism=1 -num-run-workers=2' &    
```

And then do `su www-data` and `sh /dev-tools/setup.sh database root "http://test.vipdev.lndo.site/" "VIP Dev"` I get:
````
Starting cron runner
/wp $ nohup: can't open '/home/www-data/nohup.out': Permission denied

ls -al /home/www-data/
total 8
drwxr-sr-x    2 82       82            4096 Aug 27 21:00 .
drwxr-xr-x    1 root     root          4096 Aug 27 21:00 ..
```
So it seems we ned to `chown ~/www-data` in our perm script after we change UID.

It works fine on root for me. But as mentioned above I would like to avoid that as the files created as root mapped to the host machine are not deletable by current user (e.g. the log file).

**Comment by pschoffer** [comment]
Hmm it's kinda weird... I have no idea why it keeps killing all the processes. I tried `sleep infinite &` and it also gets killed so not just because of the user rights...

**Comment by WPprodigy** [comment]
Yeah...I fumbled quite too long w/ different "ways". Can certainly redirect the logs, but I think keeping in `run.sh` for now at least is okay. The runner doesn't really care the WP isn't set up yet. Even if it takes more than 60secs, it will just fail to get the sites list on the first attempt and try again the next time. It's not super reliant on WP really, it will keep running and trying it's thing.

File: php-fpm/Dockerfile [hunk 1]
```diff
@@ -24,6 +34,8 @@ COPY php-fpm/php.ini /usr/local/etc/php/php.ini
 # move it to backup location, run-php-fpm.sh will move it to conf.d if enabled
 COPY php-fpm/xdebug.ini /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini.template
 
+COPY --from=builder /cron-runner/bin/cron-control-runner /usr/local/bin/cron-control-runner
```
Comment (pschoffer) [inline php-fpm/Dockerfile:37]: TIL Multistage builds :+1:

File: php-fpm/run.sh [hunk 2]
```diff
@@ -14,5 +14,8 @@ else
     fi
 fi
 
+# Kick off cron runner in a background process. Initial sleep is to help give WP time to be setup.
+echo "Starting cron runner"
+nohup bash -c 'sleep 60; /usr/local/bin/cron-control-runner -wp-path=/wp -wp-cli-path=/usr/local/bin/wp -get-sites-interval=120s -get-events-interval=60s -get-events-parallelism=1 -num-run-workers=2' &>/wp/log/cron-runner.log &
```
Comment (pschoffer) [inline php-fpm/run.sh:19]: Could we break up this line using  `\`  this is a bit hard to follow (scroll :) ) ?

Comment (pschoffer) [inline php-fpm/run.sh:19]: Could we use some ENV variable to be able to disable cron? OR is there any already in the go runner?

**Review by pschoffer — CHANGES_REQUESTED** [review]
Looks good left a minor nitpicks :)

**Comment by pauarge** [comment]
@WPprodigy You correctly pointed out why the build fails. I did this _hack_ to get the version number so we can tag the version accordingly. Maybe now it doesn't make that sense given that we have a Go plus a PHP version. 

We can tweak the number of lines `head` uses to get the version so the build passes, though.

File: php-fpm/run.sh (hunk 2 above)
Comment (WPprodigy) [reply php-fpm/run.sh:19]: Running this CLI on the root site will effectively stop it: https://github.com/Automattic/Cron-Control/blob/master/includes/wp-cli/class-orchestrate.php#L42-L52

Comment (pschoffer) [reply php-fpm/run.sh:19]: Can you run that with `wp` or do you need to ssh to the container?

I'm thinking in the ideal case we have `--cron=false` option when creating/updating environment. If false we would set some env variable similar to xdebug or phpmyadmin.

**Comment by WPprodigy** [comment]
I guess maybe we should consider it being called `php-fpm-with-cron`? 🙃 , then for versioning can use a specific tag of the cron runner combined w/ the php version.

**Comment by pschoffer** [comment]
I think we can just keep the php version in. Either manually or via that magic Docker parsing logic :) Right now it doesn't really matter as there is no way of choosing in dev-env which php version you want, but that might be a thing.

I see that way more likely than users changing the cron version.

**Comment by jesse-greathouse** [comment]
This branch has conflicts.

**Comment by rinatkhaziev** [comment]
So this got a bit stale :)

**Comment by WPprodigy** [comment]
I'm a bit conflicted still on shipping this as a side-effect in a container named `php-fpm`.
